usr/lib/cca_stdll/cca_specific.c

```diff
@@ -7797,6 +7797,190 @@ CK_RV token_specific_ibm_dilithium_generate_keypair(STDLL_TokData_t *tokdata,
+                    oid->keyform);
+        return CKR_KEY_SIZE_RANGE;
+    }
+
+    return CKR_OK;
+}
+
+CK_RV token_specific_ibm_dilithium_sign(STDLL_TokData_t *tokdata,
+                                        SESSION *sess,
+                                        CK_BBOOL length_only,
+                                        const struct pqc_oid *oid,
+                                        CK_BYTE *in_data,
+                                        CK_ULONG in_data_len,
+                                        CK_BYTE *signature,
+                                        CK_ULONG *signature_len,
+                                        OBJECT *key_obj)
+{
+    long return_code, reason_code, rule_array_count;
+    unsigned char rule_array[CCA_RULE_ARRAY_SIZE] = { 0, };
+    long signature_bit_length;
+    CK_ATTRIBUTE *attr;
+    CK_RV rc;
+
+    UNUSED(sess);
+
+    if (((struct cca_private_data *)tokdata->private_data)->inconsistent) {
+        TRACE_ERROR("%s\n", ock_err(ERR_DEVICE_ERROR));
+        return CKR_DEVICE_ERROR;
+    }
+
+    if (!cca_pqc_strength_supported(tokdata, CKM_IBM_DILITHIUM, oid->keyform)) {
+        TRACE_DEVEL("Dilithium keyform %lu not supported by CCA\n",
+                    oid->keyform);
+        return CKR_KEY_SIZE_RANGE;
+    }
+
+    rc = check_ibm_dilithium_data_len(oid, in_data_len);
+    if (rc != CKR_OK)
+        return rc;
+
+    if (length_only) {
+        *signature_len = CCA_MAX_DILITHIUM_SIGNATURE_LEN;
+        return CKR_OK;
+    }
+
+    if (*signature_len > CCA_MAX_DILITHIUM_SIGNATURE_LEN)
+        *signature_len = CCA_MAX_DILITHIUM_SIGNATURE_LEN;
+
+    /* Find the secure key token */
+    rc = template_attribute_get_non_empty(key_obj->template, CKA_IBM_OPAQUE,
+                                          &attr);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("Could not find CKA_IBM_OPAQUE for the key.\n");
+        return rc;
+    }
+
+    rule_array_count = 3;
+    memcpy(rule_array, "CRDL-DSAMESSAGE CRDLHASH", CCA_KEYWORD_SIZE * 3);
+
+    USE_CCA_ADAPTER_START(tokdata, return_code, reason_code)
+    RETRY_NEW_MK_BLOB_START()
+        dll_CSNDDSG(&return_code,
+                    &reason_code,
+                    NULL,
+                    NULL,
+                    &rule_array_count,
+                    rule_array,
+                    (long *)&(attr->ulValueLen),
+                    attr->pValue,
+                    (long *)&in_data_len,
+                    in_data,
+                    (long *)signature_len, &signature_bit_length, signature);
+    RETRY_NEW_MK_BLOB_END(tokdata, return_code, reason_code,
+                          attr->pValue, attr->ulValueLen)
+    USE_CCA_ADAPTER_END(tokdata, return_code, reason_code)
+
+    if (return_code != CCA_SUCCESS) {
+        TRACE_ERROR("CSNDDSG (QSA SIGN) failed. return:%ld,"
+                    " reason:%ld\n", return_code, reason_code);
+        return CKR_FUNCTION_FAILED;
+    } else if (reason_code != 0) {
```

Comment:
Are you using "0" instead of CCA_SUCCESS intentionally when comparing the reason code?

Reply:
CCA_SUCCESS is just meant for the return code, not for the reason code.... 
Could still be used there as well though....
There are hundred of similar cases, so I guess I keep it consistent with all the other places.